Source/WebCore/platform/graphics/SourceBufferPrivate.cpp

```diff
@@ -412,98 +414,104 @@ void SourceBufferPrivate::evictCodedFrames(uint64_t newDataSize, uint64_t maximu
 
     // 3. Let removal ranges equal a list of presentation time ranges that can be evicted from
     // the presentation to make room for the new data.
 
-    // NOTE: begin by removing data from the beginning of the buffered ranges, 30 seconds at
-    // a time, up to 30 seconds before currentTime.
-    MediaTime thirtySeconds = MediaTime(30, 1);
-    MediaTime maximumRangeEnd = currentTime - thirtySeconds;
+    // NOTE: begin by removing data from the beginning of the buffered ranges, timeChunk seconds at
+    // a time, up to timeChunk seconds before currentTime.
 
 #if !RELEASE_LOG_DISABLED
     uint64_t initialBufferedSize = totalTrackBufferSizeInBytes();
     DEBUG_LOG(LOGIDENTIFIER, "currentTime = ", currentTime, ", require ", initialBufferedSize + newDataSize, " bytes, maximum buffer size is ", maximumBufferSize);
 #endif
 
-    MediaTime rangeStart = MediaTime::invalidTime();
+    const auto& buffered = m_buffered->ranges();
 
-    for (auto& trackBuffer : m_trackBufferMap.values()) {
-        auto iter = trackBuffer->samples().presentationOrder().findSampleContainingOrAfterPresentationTime(MediaTime::zeroTime());
-        if (iter != trackBuffer->samples().presentationOrder().end()) {
-            MediaTime startTime = iter->first;
-            if (rangeStart.isInvalid() || startTime < rangeStart)
-                rangeStart = startTime;
-        }
-    }
-
-    if (rangeStart.isInvalid())
-        rangeStart = MediaTime::zeroTime();
+    unsigned timeChunkAsMilliseconds = evictionAlgorithmInitialTimeChunk;
+    do {
+        MediaTime timeChunk = MediaTime(timeChunkAsMilliseconds, 1000);
+        MediaTime rangeStart = buffered.minimumBufferedTime();
+        MediaTime rangeEnd = rangeStart + timeChunk;
+        MediaTime maximumRangeEnd = currentTime - timeChunk;
+        while (rangeStart < maximumRangeEnd) {
+            // 4. For each range in removal ranges, run the coded frame removal algorithm with start and
+            // end equal to the removal range start and end timestamp respectively.
+            MediaTime realRangeEnd = std::min(rangeEnd, maximumRangeEnd);
+            if (rangeStart >= realRangeEnd)
+                break;
+            removeCodedFrames(rangeStart, realRangeEnd, currentTime, isEnded);
+            isBufferFull = isBufferFullFor(newDataSize, maximumBufferSize);
+            if (!isBufferFull)
+                break;
 
-    MediaTime rangeEnd = rangeStart + thirtySeconds;
-    while (rangeStart < maximumRangeEnd) {
-        // 4. For each range in removal ranges, run the coded frame removal algorithm with start and
-        // end equal to the removal range start and end timestamp respectively.
-        removeCodedFrames(rangeStart, std::min(rangeEnd, maximumRangeEnd), currentTime, isEnded);
-        if (!isBufferFullFor(newDataSize, maximumBufferSize)) {
-            break;
+            MediaTime newRangeStart = buffered.minimumBufferedTime();
+            if (newRangeStart == rangeStart)
```

Comment:
Depending on the iterations and how small the partition is, specially if you make the `removeCodedFrames` algorithm more conservative, this can happen. In that case, you want to bail out and continue to the next partition.